update API

al-niessner opened the pull request:
## 🗒️ Summary
Changed the endpoints to match the desires of the review board. Implement all of the requested endpoints.

## ⚙️ Test Data and/or Report
Output from ut/regress.sh

## ♻️ Related Issues
Closes NASA-PDS/pds-api#202

Commit:
update API

Added all of the new endpoints in groups classes, references, and products.

Added all of the old endpoints in the group deprecrated.

Commit:
save this spot

Commit:
add more products

Commit:
changed to fit proposal

Update swagger.yml to match the proposal plus any additions and clarification on the original ticket NASA-PDS/pds-api#202.

Re-arranged the code to fit the new API. Required some new exceptions and tweaks to the basic interfaces to support the membership concept outside of just referencing in general and that a lidvid necessarily match a particular product type.

Added new swagger transmutation layers to support the easy removal of deprecated code in the future and to support easier maintenance in the future. Left the old referencing suff only because there is a chance it may come back in the newish future to support some other use cases.

The new swagger transmutation layers still all funnel through a single processor once all of the inputs have been collected. The deprecated code mostly works itself through the new API showing that what could be done still can using the inew API. Did not want to update the standard handler to also check that desired class matches the lid(vid) so did not route those but they could also be done with a touch more code change just for that demonstration whcih is why it was not done.

al-niessner commented on service/src/main/java/gov/nasa/pds/api/registry/view/ClassesSerializer.java:
```diff
@@ -0,0 +1,40 @@
+			throws IOException, HttpMessageNotWritableException {
+		List<String> quoted = new ArrayList<String>(t.size());
+        OutputStream os = outputMessage.getBody();
+        OutputStreamWriter wr = new OutputStreamWriter(os);
```
fixed

Commit:
fixed sonatypes

Commit:
fill out verify()

Given the new endpoints of /classes/{class}/{identifier} have to verify the {identifier} is actually of the type {class}. Previous commit had a blank for this function but need it filled out to test it.

jordanpadams commented on model/swagger.yml:
```diff
@@ -37,14 +61,218 @@ paths:
         - $ref: "#/components/parameters/Query"
         - $ref: "#/components/parameters/Sort"
         - $ref: "#/components/parameters/Start"
+  /classes/{class}/{identifier}/members:
```
@al-niessner do we want to support `/classes/{class}/{identifier}` as well?

al-niessner replied:
@jordanpadams 

I can add it if you like. Just out of curiosity, how is this different than /products/{identifier}?

jordanpadams commented on model/swagger.yml:
```diff
@@ -37,14 +61,218 @@ paths:
+      - $ref: "#/components/parameters/Limit"
+      - $ref: "#/components/parameters/Sort"
+      - $ref: "#/components/parameters/Start"
+  /classes/{class}/{identifier}/member-of/member-of/{versions}:
```
Is there somewhere in the API docs we could define the value-set for `{versions}`?

jordanpadams replied:
@al-niessner ☝️

al-niessner replied:
@jordanpadams

values for {versions} is defined by the yaml class Versions. To find it easily, look for Versions: and you will see that it maps to {versions} via its name and the enum choices. The docs should show it correctly as all or latest.

Because you are going to want it, it cannot be blank. It is why you have to have two nearly identical endpoints of /products/{identifier}/members and /products/{identifier}/members/{versions}.

/products/{identifier} cannot use {versions} because the all/latest changes the output between Plural/Singular respectively - plural and singular are known a priori. In the members and members-of cases they are always Plural - count can only be known a posteriori - so {versions} works.

jordanpadams reviewed (approved):
A couple questions, but approved with a pre-req that tests pass successfully.

al-niessner commented:
@jordanpadams @jimmie @tloubrieu-jpl 

Struggling right now because none of the endpoints are reachable despite no errors during build. Changed all version numbers to make sure that is not the problem, but seems not be. Trying to get it fixed and tested during driving breaks.

al-niessner commented:
> @jordanpadams @jimmie @tloubrieu-jpl
> 
> Struggling right now because none of the endpoints are reachable despite no errors during build. Changed all version numbers to make sure that is not the problem, but seems not be. Trying to get it fixed and tested during driving breaks.

@jordanpadams @jimmie @tloubrieu-jpl 

Figured it out. A limitation of swagger/spring forces all of the methods to be in the @Controller rather than inherited. Luckily eclipse put the methods in for and forwarded them to the inherited methods. Moving forward.

Commit:
methods need to be in controller

Seems there is a limitation in spring (maybe swagger) that the methods to be used in a @Controller must be present in the class and not inherited. This makes the code a bit more repetitive, but it works.

Commit:
fix a typo

Commit:
fix group problem

Commit:
adjusted regress

Had to update the endpoints to match the new API. Simply tested what already existed and did not add any new tests for new corner cases.

Had to update the number of collections and products too. Seems my latest install of data from harvest is smaller than before. However it all fits the current data in my opensearch now.

al-niessner commented:
@jordanpadams @jimmie @tloubrieu-jpl 

All working once again:

```
niessner@elysium:~/Projects/PDS/registry-api$ service/ut/regress.sh 
============================= test session starts ==============================
platform linux -- Python 3.8.10, pytest-4.6.9, py-1.8.1, pluggy-0.13.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/niessner/Projects/PDS/registry-api
collected 30 items                                                             

service/ut/base_ut.py::test_bad_group PASSED                             [  3%]
service/ut/base_ut.py::test_bad_lidvid PASSED                            [  6%]
service/ut/base_ut.py::TestAny::test_bundles PASSED                      [ 10%]
service/ut/base_ut.py::TestAny::test_collections PASSED                  [ 13%]
service/ut/base_ut.py::TestAny::test_lidvid PASSED                       [ 16%]
service/ut/base_ut.py::TestAny::test_lidvid_all PASSED                   [ 20%]
service/ut/base_ut.py::TestAny::test_lidvid_latest PASSED                [ 23%]
service/ut/base_ut.py::TestAny::test_products PASSED                     [ 26%]
service/ut/base_ut.py::TestBundles::test_bundles PASSED                  [ 30%]
service/ut/base_ut.py::TestBundles::test_collections PASSED              [ 33%]
service/ut/base_ut.py::TestBundles::test_collections_all PASSED          [ 36%]
service/ut/base_ut.py::TestBundles::test_collections_latest PASSED       [ 40%]
service/ut/base_ut.py::TestBundles::test_lidvid PASSED                   [ 43%]
service/ut/base_ut.py::TestBundles::test_lidvid_all PASSED               [ 46%]
service/ut/base_ut.py::TestBundles::test_lidvid_latest PASSED            [ 50%]
service/ut/base_ut.py::TestBundles::test_products PASSED                 [ 53%]
service/ut/base_ut.py::TestCollections::test_bundles PASSED              [ 56%]
service/ut/base_ut.py::TestCollections::test_collections PASSED          [ 60%]
service/ut/base_ut.py::TestCollections::test_lidvid PASSED               [ 63%]
service/ut/base_ut.py::TestCollections::test_lidvid_all PASSED           [ 66%]
service/ut/base_ut.py::TestCollections::test_lidvid_latest PASSED        [ 70%]
service/ut/base_ut.py::TestCollections::test_products PASSED             [ 73%]
service/ut/base_ut.py::TestCollections::test_products_all PASSED         [ 76%]
service/ut/base_ut.py::TestCollections::test_products_latest PASSED      [ 80%]
service/ut/base_ut.py::TestProducts::test_bundles PASSED                 [ 83%]
service/ut/base_ut.py::TestProducts::test_collections PASSED             [ 86%]
service/ut/base_ut.py::TestProducts::test_lidvid PASSED                  [ 90%]
service/ut/base_ut.py::TestProducts::test_lidvid_all PASSED              [ 93%]
service/ut/base_ut.py::TestProducts::test_lidvid_latest PASSED           [ 96%]
service/ut/base_ut.py::TestProducts::test_products PASSED                [100%]

========================== 30 passed in 0.81 seconds ===========================
```
